_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,440 @@
+
+Even tough the test cases are reduced to a minimum, the readability suffers a lot. Besides, we must highly pay 
+attention to not break any other test cases. Like so, we avoid interfering other tests by overriding the setup.
+
+Avoid `setUp()` methods like this:
+
+```java
+  @BeforeEach
+  void setUp() {
+    expected = createCity();
+    cityRepository = Mockito.mock(CityRepository.class);
+    cityService = new CityServiceImpl(cityRepository);
+
+    // Avoid such complex declarations
+    Mockito.when(cityRepository.save(expected))
+        .thenReturn(Optional.of(expected));
+    Mockito.when(cityRepository.find(expected.getId()))
+        .thenReturn(Optional.of(expected));
+    Mockito.when(cityRepository.findByName(expected.getName()))
+        .thenReturn(Optional.of(expected));
+    Mockito.when(cityRepository.findAllByCanton(expected.getCanton()))
+        .thenReturn(Collections.singleton(expected));
+    Mockito.when(cityRepository.findAllByCountry(expected.getCanton().getCountry()))
+        .thenReturn(Collections.singleton(expected));
+  }
+```
+
+To get simple test cases like this:
+
+```java
+  @Test
+  void save() throws ElementNotFoundException {
+    ReflectionAssert.assertReflectionEquals(expected, cityService.save(expected));
+  }
+
+  @Test
+  void find() throws ElementNotFoundException {
+    ReflectionAssert.assertReflectionEquals(expected, cityService.find(expected.getId()));
+  }
+
+  @Test
+  void delete() throws ElementNotFoundException {
+    cityService.delete(expected);
+    Mockito.verify(cityRepository).delete(expected);
+  }
+
+  @Test
+  void findByName() throws ElementNotFoundException {
+    ReflectionAssert.assertReflectionEquals(expected, cityService.findByName(expected.getName()));
+  }
+```
+
+### Don't Recycle Mocks
+
+Having a complex scenario might entrap us to recycle our mocks. However, this will lead us sooner or later to unexpected
+behaviour of our mocks. Each scenario shall be well reflected and set up and in case we need the same setup twice,
+initialize it by calling a method.  It's better to create new mocks for each test case. Otherwise, we might experience 
+unexpected behavior through interfering declarations.
+
+```java
+  void initializeScenario() {
+    // Mockito behaviour declarations
+  }
+
+  @BeforeEach
+  void setUp() {
+    MockitoAnnotations.openMocks(this);
+    cityService = new CityServiceImpl(cityRepository);
+  }
+
+  @Test
+  void testOne() {
+    initializeScenario();
+    // Test Case One
+  }
+  
+  @Test
+  void testTwo() {
+    initializeScenario();
+    // Test Case Two
+  }
+```
+
+### Write Test Cases Independent
+
+Don't expected test cases to be always executed in the same order! Write everything that belongs to our test case into
+the test method, so that a test case can be executed alone in our IDE or all together within our CI!
+This mistake often comes together with recycled mocks. At least now our alarm bell should ring, and it's time to
```

Comment:
Why does it come together with recycled mocks? 

Why is `reset()` considered a code smell?